.github/workflows/docker.yml

```diff
@@ -0,0 +1,79 @@
+name: "Build and push docker image"
+
+on:
+  workflow_call:
+    inputs:
+      tag:
+        required: true
+        description: The docker tag to use.
+        type: string
+
+jobs:
+  # We build the binaries outside of docker because here we can parallelize.
+  build_docker_image:
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
+      packages: write
+    strategy:
+      matrix:
+        include:
+          - bin: taker
+            target: x86_64-unknown-linux-gnu
+            os: ubuntu-latest
+            docker_arch: amd64
+          - bin: taker
+            target: aarch64-unknown-linux-gnu
+            os: ubuntu-latest
+            docker_arch: arm64
+          - bin: maker
+            target: x86_64-unknown-linux-gnu
+            os: ubuntu-latest
+            docker_arch: amd64
+          - bin: maker
+            target: aarch64-unknown-linux-gnu
+            os: ubuntu-latest
+            docker_arch: arm64
+    steps:
+      - uses: actions/checkout@v2.4.0
+
+      - uses: ./.github/actions/build-release-binary
+        id: build
+        with:
+          target: ${{ matrix.target }}
+          binary: ${{ matrix.bin }}
+
+      - name: Login into github registry
+        uses: docker/login-action@v1.11.0
+        with:
+          registry: ghcr.io
+          username: ${{ github.repository }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Build docker image
+        run: |
+          docker build --platform linux/${{ matrix.docker_arch }} --pull --build-arg BINARY_PATH=./target/${{ matrix.target }}/release/${{ matrix.bin }} --tag ghcr.io/${{ github.repository }}/${{matrix.bin}}:${{ inputs.tag }}-${{ matrix.docker_arch }} .
+          docker push ghcr.io/${{ github.repository }}/${{matrix.bin}}:${{ inputs.tag }}-${{ matrix.docker_arch }}
```

Comment:
I'm not 100% convinced about the approach of building a separate docker image for each binary and then combine it using manifest. 

What's the value we get from this? I don't think it is speed. Because you still have to wait for all images to build before you can combine them using the `create_manifest_image` step. 
Building the docker images using buildx is quite fast and I never thought of this being a bottleneck. The only thing we do is copy the binary into the container, and this does not take much time. 

The problem I see is that we might have half-backed images out there. E.g. building `amd64` works but for `arm64` fails. Then we have one container image `maker-amd64` but not the one for `arm64` and not the combined one.
Are we happy with this?

Reply:
> What's the value we get from this? I don't think it is speed. Because you still have to wait for all images to build before you can combine them using the `create_manifest_image` step.
> Building the docker images using buildx is quite fast and I never thought of this being a bottleneck. The only thing we do is copy the binary into the container, and this does not take much time.

Correct! I did not optimize for speed but for simplicity!

- The previous approach was tricky to understand because we had to copy the binaries into a particular directory structure to allow `buildx`'s internall parallelism to work.
- This also leaked into the Dockerfile definition where it was unclear, what these build parameters do.
- The buildx approach also depended on being able to download the binaries from the _preview_ release which was a legacy concept that we were no longer actively utilizing and had its own tech-debt built up. This could be mitigated by using the `actions/{upload,download}-artifact` actions but that alone also adds complexity to the workflow again.

I think the buildx approach is really powerful and simple if the entire build process is within the Dockerfile. Then building a multi-arch image becomes a one-liner. We don't do that for performance reasons so it IMO loses a lot of its appeal.

We already have a build-matrix for building all the different binaries so it felt natural to me to just tuck the building of the specific container on top of that. The convenience of a multi-arch image is then just something we add on top, once all the pipelines are done!

> The problem I see is that we might have half-backed images out there. E.g. building `amd64` works but for `arm64` fails. Then we have one container image `maker-amd64` but not the one for `arm64` and not the combined one.
> Are we happy with this?

I am obviously biased towards the proposed solution but also objectively considered, I don't think it is an actual issue. We can delete containers if we want or we trigger the pipeline again if it was a ephemeral error.

Reply:
Imho we exchanged one complexity for a different one, but that's fine with me.
I'm happy to go with your proposal and if we ever have partially built docker images just have to clean them up manually 👍

Reply:
> Imho we exchanged one complexity for a different one, but that's fine with me.

That is probably true.

> I'm happy to go with your proposal

Cool, thanks! What is being proposed here makes releases a bit easier because we don't need to manually trigger the building of docker images for a release after they have been attached so I'd say it is a net-positive overall. We can always iterate on the implementation details of how the containers are built.